File: policy-engine/docs/research/policy-operations/register-closure/task-l/locale_census.mjs
#!/usr/bin/env node
/** Node recursive-descent implementation of the complete INT-R6 locale census. */

import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";

const SCHEMA_VERSION = "policyos.research.task_l.locale_leaf_census.v1";
const DECODER_ID = "node.recursive_descent.v1";

class ParseFailure extends Error {
  constructor(code, detail) {
    super(`${code}: ${detail}`);
    this.code = code;
    this.detail = detail;
  }
}

function pointer(parts) {
  return `/${parts.map((part) => part.replaceAll("~", "~0").replaceAll("/", "~1")).join("/")}`;
}

function decodeDocument(text, fileName) {
  let offset = 0;

  function fail(code, detail = `offset=${offset}`) {
    throw new ParseFailure(code, `${fileName}:${detail}`);
  }

  function skipWhitespace() {
    while (offset < text.length && /[\t\n\r ]/.test(text[offset])) offset += 1;
  }

  function parseString() {
    if (text[offset] !== '"') fail("node_recursive_string_expected");
    offset += 1;
    let result = "";
    while (offset < text.length) {
      const current = text[offset];
      offset += 1;
      if (current === '"') return result;
      if (current === "\\") {
        if (offset >= text.length) fail("node_recursive_escape_truncated");
        const escaped = text[offset];
        offset += 1;
        const simple = {
          '"': '"',
          "\\": "\\",
          "/": "/",
          b: "\b",
          f: "\f",
          n: "\n",
          r: "\r",
          t: "\t",
        };
        if (Object.hasOwn(simple, escaped)) {
          result += simple[escaped];
          continue;
        }
        if (escaped !== "u") fail("node_recursive_escape_invalid", `offset=${offset - 1}`);
        const hexadecimal = text.slice(offset, offset + 4);
        if (!/^[0-9a-fA-F]{4}$/.test(hexadecimal)) {
          fail("node_recursive_unicode_escape_invalid", `offset=${offset}`);
        }
        offset += 4;
        const high = Number.parseInt(hexadecimal, 16);
        if (
          high >= 0xd800 &&
          high <= 0xdbff &&
          text.slice(offset, offset + 2) === "\\u" &&
          /^[0-9a-fA-F]{4}$/.test(text.slice(offset + 2, offset + 6))
        ) {
          const low = Number.parseInt(text.slice(offset + 2, offset + 6), 16);
          if (low >= 0xdc00 && low <= 0xdfff) {
            result += String.fromCodePoint(0x10000 + ((high - 0xd800) << 10) + low - 0xdc00);
            offset += 6;
            continue;
          }
        }
        result += String.fromCharCode(high);
        continue;
      }
      if (current.charCodeAt(0) < 0x20) {
        fail("node_recursive_control_character", `offset=${offset - 1}`);
      }
      result += current;
    }
    fail("node_recursive_string_unterminated");
  }

  function parseNumber() {
    const remaining = text.slice(offset);
    const match = remaining.match(/^-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?/);
    if (match === null) fail("node_recursive_number_invalid");
    offset += match[0].length;
    const result = Number(match[0]);
    if (!Number.isFinite(result)) fail("node_recursive_number_non_finite");
    return result;
  }

  function parseArray(parts) {
    offset += 1;
    const result = [];
    skipWhitespace();
    if (text[offset] === "]") {
      offset += 1;
      return result;
    }
    while (offset < text.length) {
      result.push(parseValue([...parts, String(result.length)]));
      skipWhitespace();
      if (text[offset] === "]") {
        offset += 1;
        return result;
      }
      if (text[offset] !== ",") fail("node_recursive_array_separator_missing");
      offset += 1;
      skipWhitespace();
    }
    fail("node_recursive_array_unterminated");
  }

  function parseObject(parts) {
    offset += 1;
    const result = Object.create(null);
    const keys = new Set();
    skipWhitespace();
    if (text[offset] === "}") {
      offset += 1;
      return result;
    }
    while (offset < text.length) {
      if (text[offset] !== '"') fail("node_recursive_object_key_expected");
      const key = parseString();
      if (keys.has(key)) {
        throw new ParseFailure("node_recursive_duplicate_key", `${fileName}:${pointer([...parts, key])}`);
      }
      keys.add(key);
      skipWhitespace();
      if (text[offset] !== ":") fail("node_recursive_object_colon_missing");
      offset += 1;
      skipWhitespace();
      result[key] = parseValue([...parts, key]);
      skipWhitespace();
      if (text[offset] === "}") {
        offset += 1;
        return result;
      }
      if (text[offset] !== ",") fail("node_recursive_object_separator_missing");
      offset += 1;
      skipWhitespace();
    }
    fail("node_recursive_object_unterminated");
  }

  function parseValue(parts) {
    skipWhitespace();
    const current = text[offset];
    if (current === "{") return parseObject(parts);
    if (current === "[") return parseArray(parts);
    if (current === '"') return parseString();
    if (text.startsWith("true", offset)) {
      offset += 4;
      return true;
    }
    if (text.startsWith("false", offset)) {
      offset += 5;
      return false;
    }
    if (text.startsWith("null", offset)) {
      offset += 4;
      return null;
    }
    if (current === "-" || /[0-9]/.test(current ?? "")) return parseNumber();
    fail("node_recursive_value_invalid", `${pointer(parts)}:offset=${offset}`);
  }

  const result = parseValue([]);
  skipWhitespace();
  if (offset !== text.length) fail("node_recursive_trailing_content");
  return result;
}

function flatten(value, parts = [], output = new Map()) {
  if (Array.isArray(value)) {
    value.forEach((item, index) => flatten(item, [...parts, String(index)], output));
    return output;
  }
  if (value !== null && typeof value === "object") {
    Object.keys(value)
      .sort()
      .forEach((key) => flatten(value[key], [...parts, key], output));
    return output;
  }
  if (typeof value !== "string") {
    throw new ParseFailure("node_recursive_non_string_leaf", `${pointer(parts)}:${typeof value}`);
  }
  output.set(pointer(parts), value);
  return output;
}

function digestRows(rows) {
  const body = rows.map((row) => JSON.stringify(row)).join("\n");
  return `sha256:${crypto.createHash("sha256").update(body, "utf8").digest("hex")}`;
}

function buildReport(directoryArgument) {
  const directory = path.resolve(directoryArgument);
  let entries;
  try {
    entries = fs
      .readdirSync(directory, { withFileTypes: true })
      .sort((left, right) => (left.name < right.name ? -1 : left.name > right.name ? 1 : 0));
  } catch (error) {
    throw new ParseFailure("node_locale_directory_unreadable", error.constructor.name);
  }
  const unexpected = entries
    .filter((entry) => !entry.isFile() || path.extname(entry.name) !== ".json")
    .map((entry) => entry.name);
  if (unexpected.length > 0) {
    throw new ParseFailure("node_locale_entry_ambiguous", JSON.stringify(unexpected));
  }
  if (entries.length === 0) throw new ParseFailure("node_locale_directory_empty", directory);

  const leaves = new Map();
  for (const entry of entries) {
    const locale = path.basename(entry.name, ".json");
    if (leaves.has(locale)) throw new ParseFailure("node_locale_identity_duplicate", locale);
    let text;
    try {
      text = fs.readFileSync(path.join(directory, entry.name), "utf8");
    } catch (error) {
      throw new ParseFailure("node_locale_unreadable", `${entry.name}:${error.constructor.name}`);
    }
    const decoded = decodeDocument(text, entry.name);
    if (decoded === null || Array.isArray(decoded) || typeof decoded !== "object") {
      throw new ParseFailure("node_locale_root_not_object", entry.name);
    }
    leaves.set(locale, flatten(decoded));
  }
  if (!leaves.has("en")) throw new ParseFailure("node_reference_locale_missing", "en");

  const allPathSets = [...leaves.values()].map((localeLeaves) => new Set(localeLeaves.keys()));
  const union = new Set(allPathSets.flatMap((paths) => [...paths]));
  const intersection = new Set(
    [...allPathSets[0]].filter((leafPath) => allPathSets.every((paths) => paths.has(leafPath))),
  );
  const localeReports = Object.create(null);
  for (const locale of [...leaves.keys()].sort()) {
    const rows = [...leaves.get(locale).entries()].sort(([left], [right]) =>
      left < right ? -1 : left > right ? 1 : 0,
    );
    localeReports[locale] = {
      leaf_count: rows.length,
      path_digest: digestRows(rows.map(([leafPath]) => [leafPath, ""])),
      path_value_digest: digestRows(rows),
    };
  }

  const english = leaves.get("en");
  const comparisons = Object.create(null);
  for (const locale of [...leaves.keys()].sort()) {
    if (locale === "en") continue;
    const target = leaves.get(locale);
    const common = [...english.keys()].filter((leafPath) => target.has(leafPath)).sort();
    comparisons[locale] = {
      common_leaf_count: common.length,
      missing_from_target_count: [...english.keys()].filter((leafPath) => !target.has(leafPath)).length,
      target_only_count: [...target.keys()].filter((leafPath) => !english.has(leafPath)).length,
      identical_value_count: common.filter((leafPath) => english.get(leafPath) === target.get(leafPath)).length,
      different_value_count: common.filter((leafPath) => english.get(leafPath) !== target.get(leafPath)).length,
    };
  }

  return {
    schema_version: SCHEMA_VERSION,
    decoder_id: DECODER_ID,
    decoder_implementation: "standalone Node recursive-descent decoder",
    directory_files: entries.map((entry) => entry.name),
    directory_file_count: entries.length,
    union_leaf_count: union.size,
    intersection_leaf_count: intersection.size,
    locales: localeReports,
    comparisons_to_en: comparisons,
  };
}

function main(arguments_) {
  if (arguments_.length !== 1) {
    process.stderr.write("usage: locale_census.mjs LOCALE_DIRECTORY\n");
    return 2;
  }
  try {
    const report = buildReport(arguments_[0]);
    process.stdout.write(`${JSON.stringify(report)}\n`);
    return 0;
  } catch (error) {
    if (error instanceof ParseFailure) {
      process.stderr.write(`${error.code}: ${error.detail}\n`);
      return 2;
    }
    throw error;
  }
}

process.exitCode = main(process.argv.slice(2));

File: policy-engine/docs/research/policy-operations/register-closure/task-l/locale_census.py
#!/usr/bin/env python3
"""Python implementation of the complete INT-R6 locale leaf census."""

from __future__ import annotations

import hashlib
import json
import sys
from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Mapping, Sequence


SCHEMA_VERSION = "policyos.research.task_l.locale_leaf_census.v1"
DECODER_ID = "python.json.object_pairs_hook"


class LocaleCensusError(ValueError):
    """Raised when a locale member is unreadable or structurally ambiguous."""

    def __init__(self, code: str, detail: str) -> None:
        super().__init__(f"{code}: {detail}")
        self.code = code
        self.detail = detail


class LocaleCensusMismatchError(AssertionError):
    """Raised when two independently produced census reports disagree."""

    def __init__(self, code: str, detail: str) -> None:
        super().__init__(f"{code}: {detail}")
        self.code = code
        self.detail = detail


def _unique_object(pairs: Sequence[tuple[str, object]]) -> dict[str, object]:
    result: dict[str, object] = {}
    for key, value in pairs:
        if key in result:
            raise LocaleCensusError("python_json_duplicate_key", key)
        result[key] = value
    return result


def _pointer(parts: tuple[str, ...]) -> str:
    return "/" + "/".join(part.replace("~", "~0").replace("/", "~1") for part in parts)


def _flatten(value: object, *, parts: tuple[str, ...] = ()) -> dict[str, str]:
    if isinstance(value, dict):
        flattened: dict[str, str] = {}
        for key in sorted(value):
            if not isinstance(key, str):
                raise LocaleCensusError("python_json_non_string_key", _pointer(parts))
            flattened.update(_flatten(value[key], parts=(*parts, key)))
        return flattened
    if isinstance(value, list):
        flattened = {}
        for index, item in enumerate(value):
            flattened.update(_flatten(item, parts=(*parts, str(index))))
        return flattened
    if not isinstance(value, str):
        raise LocaleCensusError(
            "python_json_non_string_leaf",
            f"{_pointer(parts)}:{type(value).__name__}",
        )
    return {_pointer(parts): value}


def _digest_rows(rows: Sequence[tuple[str, str]]) -> str:
    encoded = "\n".join(
        json.dumps([path, value], ensure_ascii=False, separators=(",", ":")) for path, value in rows
    ).encode("utf-8")
    return "sha256:" + hashlib.sha256(encoded).hexdigest()


def _parse_locale(path: Path) -> dict[str, str]:
    try:
        text = path.read_text(encoding="utf-8")
    except (OSError, UnicodeError) as exc:
        raise LocaleCensusError(
            "python_locale_unreadable", f"{path.name}:{type(exc).__name__}"
        ) from exc
    try:
        parsed = json.loads(text, object_pairs_hook=_unique_object)
    except LocaleCensusError:
        raise
    except json.JSONDecodeError as exc:
        raise LocaleCensusError(
            "python_json_decode_error", f"{path.name}:{exc.lineno}:{exc.colno}"
        ) from exc
    if not isinstance(parsed, dict):
        raise LocaleCensusError("python_locale_root_not_object", path.name)
    return _flatten(parsed)


def build_report(directory: Path) -> dict[str, object]:
    """Parse every directory member and return a complete identity census."""

    directory = directory.resolve()
    try:
        entries = sorted(directory.iterdir(), key=lambda path: path.name)
    except OSError as exc:
        raise LocaleCensusError("python_locale_directory_unreadable", type(exc).__name__) from exc
    unexpected = [path.name for path in entries if not path.is_file() or path.suffix != ".json"]
    if unexpected:
        raise LocaleCensusError("python_locale_entry_ambiguous", repr(unexpected))
    if not entries:
        raise LocaleCensusError("python_locale_directory_empty", directory.as_posix())

    leaves: dict[str, dict[str, str]] = {}
    for path in entries:
        locale = path.stem
        if locale in leaves:
            raise LocaleCensusError("python_locale_identity_duplicate", locale)
        leaves[locale] = _parse_locale(path)
    if "en" not in leaves:
        raise LocaleCensusError("python_reference_locale_missing", "en")

    all_path_sets = [set(locale_leaves) for locale_leaves in leaves.values()]
    union = set().union(*all_path_sets)
    intersection = set.intersection(*all_path_sets)
    locale_reports: dict[str, dict[str, object]] = {}
    for locale, locale_leaves in sorted(leaves.items()):
        rows = sorted(locale_leaves.items())
        path_rows = [(path, "") for path, _ in rows]
        locale_reports[locale] = {
            "leaf_count": len(rows),
            "path_digest": _digest_rows(path_rows),
            "path_value_digest": _digest_rows(rows),
        }

    english = leaves["en"]
    comparisons: dict[str, dict[str, int]] = {}
    for locale, target in sorted(leaves.items()):
        if locale == "en":
            continue
        common = sorted(set(english) & set(target))
        comparisons[locale] = {
            "common_leaf_count": len(common),
            "missing_from_target_count": len(set(english) - set(target)),
            "target_only_count": len(set(target) - set(english)),
            "identical_value_count": sum(english[path] == target[path] for path in common),
            "different_value_count": sum(english[path] != target[path] for path in common),
        }

    return {
        "schema_version": SCHEMA_VERSION,
        "decoder_id": DECODER_ID,
        "decoder_implementation": "stdlib json with duplicate-key object-pairs hook",
        "directory_files": [path.name for path in entries],
        "directory_file_count": len(entries),
        "union_leaf_count": len(union),
        "intersection_leaf_count": len(intersection),
        "locales": locale_reports,
        "comparisons_to_en": comparisons,
    }


def _semantic_report(report: Mapping[str, object]) -> dict[str, object]:
    return {
        key: copy_value
        for key, copy_value in report.items()
        if key not in {"decoder_id", "decoder_implementation"}
    }


def reconcile_reports(
    first: Mapping[str, object], second: Mapping[str, object]
) -> dict[str, object]:
    """Require exact agreement while preserving distinct decoder provenance."""

    if first.get("decoder_id") == second.get("decoder_id"):
        raise LocaleCensusMismatchError(
            "decoder_independence_not_established", repr(first.get("decoder_id"))
        )
    first_semantic = _semantic_report(first)
    second_semantic = _semantic_report(second)
    if first_semantic != second_semantic:
        raise LocaleCensusMismatchError(
            "parser_reports_disagree",
            json.dumps(
                {"first": first_semantic, "second": second_semantic},
                ensure_ascii=False,
                sort_keys=True,
                separators=(",", ":"),
            ),
        )
    canonical = json.dumps(
        first_semantic,
        ensure_ascii=False,
        sort_keys=True,
        separators=(",", ":"),
    ).encode("utf-8")
    return {
        "status": "independently_reconciled",
        "decoder_ids": [first.get("decoder_id"), second.get("decoder_id")],
        "directory_file_count": first_semantic["directory_file_count"],
        "census_digest": "sha256:" + hashlib.sha256(canonical).hexdigest(),
    }


def main(argv: Sequence[str] | None = None) -> int:
    """Print the census as canonical-key JSON; ambiguous input exits two."""

    arguments = list(sys.argv[1:] if argv is None else argv)
    if len(arguments) != 1:
        print("usage: locale_census.py LOCALE_DIRECTORY", file=sys.stderr)  # noqa: T201
        return 2
    try:
        report = build_report(Path(arguments[0]))
    except LocaleCensusError as exc:
        print(f"{exc.code}: {exc.detail}", file=sys.stderr)  # noqa: T201
        return 2
    print(  # noqa: T201
        json.dumps(report, ensure_ascii=False, sort_keys=True, separators=(",", ":"))
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
